src/pTSL.py

```diff
@@ -1,36 +1,263 @@
+    '''
+    def val_corpus(self, corpus_file):
+        return self.val_corpus_helper(corpus_file)
+
+    def val_corpus_helper(self, corpus_file):
+        # read file into data
+        corpus_data = self.read_corpus_file(corpus_file)
+        cond_probs = []
+        input_strings = []
+        for string in corpus_data:
+            cond_probs.append(self.val(string))
+            input_strings.append(' '.join(map(str, string)))
+        return dict(zip(input_strings, cond_probs))
+
+    def print_conditional_probabilities(self, corpus_file, verbose=False):
+        # read file into data
+        corpus_data = self.read_corpus_file(corpus_file)
+        for index, string in enumerate(corpus_data):
+            projections = self.get_projections(string)
+            result = self.val_helper(projections)
+            print("Input {}: {}".format(index+1, ' '.join(map(str, string))))
+            print("Input {} Probability: {}".format(index+1, round(result, 4)))
+            if verbose:
+                print("\nProjection Probabilities for input {}".format(index+1))
+                for sub, value in projections:
+                    if not value:
+                        continue
+                    print([' '.join(map(str, sub))], ': ', round(value, 3))
+                print()
+
+    def evaluate_proj(self, random_proj, param, corpus_probs):
+        # return sum of sqaured errors
+
+        for i, t in enumerate(param):
+            self.probs[t] = random_proj[i]
+
+        sse = 0
+        for string, y in corpus_probs:
+            sse += (self.val(string) - y)**2
+
+        return sse
+
+    def train(self, corpus_file):
+        # [[['a','b'],1],[]]
+        corpus_probs = self.read_corpus_file(corpus_file, True)
+        # get list of tiers
+        # from initialized ptsl (self.probs and self.factor) and corpus data, create complete list of alphabet (no duplicates)
+        param = {x for sublist in corpus_probs for x in sublist[0]}
+        # remove fixed alphabets
+        param = list(param - {x for x in self.probs})
+
+        # create bounds
+        # instead of limiting bound for fixed value, I removed it completely from the parameter
+        bounds = [(0, 1) for i in range(len(param))]
+
+        # bounds = []
+        # for t in param:
+        #     if t in self.probs:
+        #         bounds.append((self.probs[t], self.probs[t]))
+        #     else:
+        #         bounds.append((0, 1))
+
+        # randomly initialize parameter - this will be the input
+        random_proj = rand(len(param))
+        # run the minimize function
+        proj_res = minimize(self.evaluate_proj,
+                            random_proj,
+                            bounds=bounds,
+                            method='L-BFGS-B',
+                            args=(param, corpus_probs))
+
+        optimal_projection = dict(zip(param, proj_res.x))
+        self.probs.update(optimal_projection)

```

Comment:
I don't think this line is strictly necessary, since you update `self.probs` on each call to `evaluate_proj`.